### PR TITLE
Build Linux releases on big managed runners

### PR DESCRIPTION
**Pull request by GMNGeoffrey**
This speed the linux builds up a bit, bringing the time for the longest
job down from ~5 hours to ~20 minutes. Note that this is *only* the
Linux jobs. The mac ones still take about 4 hours. This should still
help when iterating on the release though and for faster failure
indicators (it was indeed helpful when I was iterating here).

I ran into issues when testing because I was using a package suffix in
the workflow dispatch, which evidently had never actually been tested
because it was totally broken. This gave me a lot of wonderful
opportunity to bash my head against bash and I reworked a lot of the
`build_linux_package.sh` script. In retrospect, I wish I'd just removed
the `package_suffix` feature.

Test run: https://github.com/iree-org/iree/actions/runs/2923210349

skip-ci

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
custom managed runners evidently don't really work yet. I theoretically created them, but they seem to be impossible to target and the job just hangs indefinitely

**Comment by GMNGeoffrey** [comment]
Ok fixed up the package suffix thing. Another build: https://github.com/iree-org/iree/actions/runs/2914524943

**Comment by GMNGeoffrey** [comment]
Once more: https://github.com/iree-org/iree/actions/runs/2914625750

File: runtime/setup.py [hunk 1]
```diff
@@ -27,25 +27,26 @@
...
 import subprocess
 import sys
 import sysconfig
-
 from distutils.command.build import build as _build
-from setuptools import find_namespace_packages, setup, Extension
+from gettext import install
+from multiprocessing.spawn import prepare
+
+from setuptools import Extension, find_namespace_packages, setup
```
Comment (GMNGeoffrey) [inline runtime/setup.py:42]: Evidently I have autoformat on save turned on

File: runtime/setup.py [hunk 2]
```diff
@@ -360,7 +361,7 @@ def generate_version_py():
   custom_package_suffix = ""
 
 setup(
-    name=f"iree-runtime{PACKAGE_SUFFIX}{custom_package_suffix}",
+    name=f"iree-runtime{custom_package_suffix}{PACKAGE_SUFFIX}",
```
Comment (GMNGeoffrey) [inline runtime/setup.py:364]: The other way round we get packages like "iree_runtime_gcmn_testing_instrumented" whereas I think we want "iree_runtime_instrumented_gcmn_testing" as we refer to "iree-runtime-instrumented" as a package

**Comment by GMNGeoffrey** [comment]
This run *almost* succeeded: https://github.com/iree-org/iree/runs/8006465144

**Comment by GMNGeoffrey** [comment]
I think this is mostly ready for review

File: .github/workflows/build_package.yml [hunk 3]
```diff
@@ -27,39 +27,39 @@ on:
...
       fail-fast: false
       matrix:
         include:
           # Ubuntu packages.
-          - os: ubuntu-20.04
-            build_family: linux
-            build_package: main-dist-linux
+          - runs-on: [managed-releaser, os-family=Linux, runner-group=releaser]
+            build-family: linux
+            build-package: main-dist-linux
```
Comment (ScottTodd) [inline .github/workflows/build_package.yml:40]: Hey @powderluv , are you using this workflow downstream? I think setting these jobs to run on self-hosted runners could interfere with that.

Comment (ScottTodd) [reply .github/workflows/build_package.yml:40]: Ping, @powderluv ? I'm not sure if GitHub Actions has a way to make this configurable for downstream repositories. You might want to fork this workflow if you depend on it.

**Review by ScottTodd — APPROVED** [review]
Generally LGTM, excited for the speed-ups. Impact on downstreams has me hesitate a little, but 4 hour releases aren't worth the convenience.

**Comment by powderluv** [comment]
Thanks for the ping. Yes we use it downstream for our SHARK-runtime releases. We can carry a fork or if we want we can leave the "one shot" builds that we use to be self hosted.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 4]
```diff
@@ -37,112 +37,122 @@
...
-  output_dir="$(cd $output_dir && pwd)"
+  mkdir -p "${output_dir}"
+  output_dir="$(cd "${output_dir}" && pwd)"
   echo "Outputting to ${output_dir}"
   mkdir -p "${output_dir}"
   docker run --rm \
-    -v "${repo_root}:/main_checkout/iree" \
-    -v "${output_dir}:/wheelhouse" \
+    -v "${repo_root}:${repo_root}" \
+    -v "${output_dir}:${output_dir}" \
```
Comment (powderluv) [inline build_tools/python_deploy/build_linux_packages.sh:62]: Eventually we may want to set UID/GID so we don't leave root:root owned files around if we decide to run this on a workstation to test. I added some flags here: https://github.com/llvm/torch-mlir/pull/1234

**Review by powderluv — COMMENTED** [review]
some minor feedback.

File: build_tools/python_deploy/build_linux_packages.sh [hunk 5]
```diff
@@ -37,112 +37,122 @@
...
 this_dir="$(cd $(dirname $0) && pwd)"
 script_name="$(basename $0)"
-repo_root="$(cd $this_dir/../../ && pwd)"
-script_name="$(basename $0)"
+repo_root="$(cd "${this_dir}" && git rev-parse --show-toplevel)"
 manylinux_docker_image="${manylinux_docker_image:-gcr.io/iree-oss/manylinux2014_x86_64-release@sha256:b09c10868f846308bad2eab253a77d0a3f097816c40342bc289d8e62509bc5f9}"
 python_versions="${override_python_versions:-cp37-cp37m cp38-cp38 cp39-cp39 cp310-cp310}"
 output_dir="${output_dir:-${this_dir}/wheelhouse}"
 packages="${packages:-iree-runtime iree-runtime-instrumented iree-compiler}"
+package_suffix="${package_suffix:-}"
```
[the remark this answers is not in the document]
Comment (GMNGeoffrey) [reply build_tools/python_deploy/build_linux_packages.sh:49]: Yeah if you set them with `export` or via GitHub's `env`

File: build_tools/python_deploy/build_linux_packages.sh (hunk 4 above)
Comment (GMNGeoffrey) [reply build_tools/python_deploy/build_linux_packages.sh:62]: Yeah we do that for all our other docker stuff. This was developed totally separately and does a lot of things differently

**Comment by GMNGeoffrey** [comment]
> Thanks for the ping. Yes we use it downstream for our SHARK-runtime releases. We can carry a fork or if we want we can leave the "one shot" builds that we use to be self hosted.

Not sure I understand what you mean by the second option, but I think you might actually be able to create managed runners that match ours. The option only appears for me at the org level, for some reason.

I think the URL should be:
https://github.com/organizations/nod-ai/settings/actions/custom-hosted-runners/new

**Comment by powderluv** [comment]
oh that is a good idea to add a similar named runner. We have self-hosted runners anyway.

**Pull request merged**